KAFKA-3923: Make KafkaMetric not final, update JmxReporter and unit tests

stepio opened the pull request:
This is a patch for the previously highlighted issue:
https://issues.apache.org/jira/browse/KAFKA-3923

I've minimised the scope of changes, so that now Metrics class is not affected and generics are introduced on the method level instead of class level.

stepio commented:
It's funny that there are both FakeMetricsReporter and MockMetricsReporter - maybe the first one may be dropped.

stepio committed:
KAFKA-3923: Remove unused FakeMetricsReporter
 - MockMetricsReporter can be used instead

stepio commented:
Hello, @guozhangwang ,

May I ease the process of reviewing for you by any chance?

stepio commented:
Any updates?

guozhangwang commented:
Sorry for the late reply @stepio , I think the motivation behind this PR makes sense, but the proposed change modifies `MetricsReporter` interface, which will break anyone implementing this class (i.e. not source compatible).

I'm wondering for making the mocking easier, and I'm assuming by mocking you want to add some bookkeeping logic for the function called, etc, would it be simpler to just make `KafkaMetric` not final? cc @ijuma @junrao .

stepio commented:
Hi @guozhangwang , it's Ok. Better late than never, so thank you anyway.

Your solution is simpler and safer, still allowing easy testing without powermock "hacks". I don't think that there are many implementations of `MetricsReporter` interface (did not find any by searching on github), but still removing "final" is safer. My approach just improves a bit the architecture - for example, similar generics are already used in `Consumer` and `Producer` interfaces.

So you're the maintainers and decision is yours.

guozhangwang commented:
@stepio I am actually aware of a couple of `MetricsReporter` implementations, and some of them comes from `confluentinc` open source repos :P Personally I agree that your fix is actually more ideal, and that we did make a mistake at the very beginning when writing `MetricsReporter`. But for retaining backward compatibility we'd probably have to go with the sub-optimal approach.

I have synced with @ijuma about this, and either myself or Ismael will be happy to help review and merge this once you refactored this PR.

stepio committed:
KAFKA-3923: Remove 'final' from 'KafkaMetric'
 - add/fix related javadocs

stepio commented:
Hi @guozhangwang, hope you're fine.

I've did the requested updates, although now I'm a bit ashamed to push such a trivial change :)
As a bonus, I've did minor clean-up in interface and added the javadoc, describing approach, which saved me eventually and was not clear for me without external help.

I see some test failures, but I'd rather blame race conditions in tests... What do you think?

P.S.: I've also kept the original "major" approach, but moved it to another branch, just for history:
https://github.com/stepio/kafka/tree/feature/KAFKA-3923

guozhangwang commented:
test this please.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:
```diff
@@ -93,6 +93,7 @@
     private static final long WINDOW_SIZE = TimeUnit.MILLISECONDS.convert(2, TimeUnit.DAYS);
     private static final int NUM_PARTITIONS = 2;
     private static final int NUM_REPLICAS = NUM_BROKERS;
+    private static final int MAX_WAIT_MS = Integer.getInteger("kafka.test.max.wait.ms", 60000);
```
Any reason to increase from 30 secs to 60?

stepio replied:
In one place there already was 60, in others - 30. One of the cases when I got test failure was timeout, so I've decided to add 30 seconds to see if that helps.

Also now you're able to configure different timeout values for different environments using environmental variables - use shorter timeout for faster PC/server if necessary.

ijuma replied:
It seems a bit ad-hoc to have this environmental variable for one test only.

stepio replied:
It's an attempt to help you solving random test failures, which I see from time to time. As I understand, you have a large "park" of build-test servers with rather different configuration, so configurable timeouts may be aid with this. In the future same variable can be used in other tests as well.

But still I believe that it's a decision of maintainer whether the proposed update suits the overall solution. So waiting for your suggestions.

ijuma replied:
Making timeouts configurable could be a good idea, but it's better done in a general way in its own PR.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:
```diff
@@ -313,16 +313,15 @@ public boolean conditionMet() {
                     } catch (final IllegalStateException e) {
-                        // Kafka Streams instance may have closed but rebalance hasn't happened
-                        return false;
+                        throw new AssertionError("Kafka Streams instance may have closed but re-balance hasn't happened", e);
```
Is this change intentional? Ditto below.

stepio replied:
It's intentional - I'll explain my reasons and you tell if I'm right or wrong.

In this test, utility (test-specific) method is expected to return true, so false is an error. And here we get false in several different cases without any ability to distinguish between them. So I've decided to throw `AssertionError` with nested exception instead - this should help investigating what exactly failed.

stepio replied:
My bad, I've missed additional check in `TestUtils`, reverting this.

guozhangwang reviewed (commented):
A couple minor comments.

stepio committed:
KAFKA-3923: Trying to do something with test failures
 - increased test timeout from 300000 to 600000, made it configurable with system properties
 - throw assertion exceptions to allow more precise failure details

stepio commented:
> test this please.

Do you mean that I should write some specific tests for removal of `final`? Some tests with inner/anonymous class extending `KafkaMetric`?

stepio commented:
Updated tests... and the main code as well... Sorry, sometimes it's hard to stop :)

stepio commented:
I've done a kind of "demo" for using assertj:
https://github.com/stepio/kafka/commit/9ebe1d85d2cc772eb333cb9b7369749fb123c99b

Will this suit Kafka project? Or would you prefer staying with pure junit?

stepio commented:
Dear @guozhangwang ,

I've done the necessary changes, please take a look when you have some time

guozhangwang commented on clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:
```diff
@@ -63,69 +64,74 @@ public void configure(Map<String, ?> configs) {}
 
     /**
-     * @param metricName
+     * Calculates name of {@link KafkaMbean} instance, using the given {@link MetricName}.
```
nit: We use singular verbs in other functions (e.g. line 54 above), would be better to be consistent.

stepio replied:
Sure, thanks

guozhangwang commented on clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:
```diff
@@ -160,17 +184,31 @@ private void reregister(KafkaMbean mbean) {
+    }
+
+    protected static class KafkaMbean implements DynamicMBean {
+        private final String metricName;
```
Why do we want to remember the metricName locally?

stepio replied:
It allows getting the appropriate `KafkaMBean` from `mbeans` map. But now I think that better approach would have been to update the map itself:
```java
    private final Map<ObjectName, KafkaMbean> mbeans = new HashMap<ObjectName, KafkaMbean>();
```
What do you think?

guozhangwang replied:
How about just use `metric.metricName` instead of `mbean.metricName` since in all its callers the `metric` object is still available?

stepio replied:
It's impossible, as same `KafkaMbean` refers to several different `MetricName` objects, so actual key is calculated with a separate method:
```java
    protected String getMBeanName(MetricName metricName) {
...
    }
```
Basically it gets just `group` and `tags`, ignoring `name`. So I'm able to get rid of that questionable extra variable only with extra calls to `getMBeanName`.

In fact, I'd prefer to do next updates:
1. Update the map as described above.
2. Hide `getMBeanName` into `KafkaMBean`.
3. Construct `KafkaMBean` with `MetricName` instead of `String`.

guozhangwang commented on clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java:
```diff
@@ -16,25 +16,123 @@
+        }
+    }
+
+    private static class MockKafkaMetric extends KafkaMetric {
```
This mock seems not much different from the extended `KafkaMetric`?

stepio replied:
Sure... It just shows that extending works correctly... Any suggestions to make it better?

guozhangwang replied:
I think your initial motivation of this PR is that you have a concrete test case where you want to use some mocks of `KafkaMetric`, if you think it is general enough we can add it as a separate class in o.a.k.test.tests.../common/metrics.

stepio replied:
The initial motivation was that i could not create an instance of `KafkaMetric` in tests. I've just needed an object to pass it into my `MetricsReporter` implementation.
And basically my mock resolves it - I got a static "fabric" method `of` there: now I'm able to isolate `MetricsReporter` implementation from `Metrics` and test just the case I need.

Also, when I've raised a question on StackOverFlow, @artembilan from Pivotal showed me another approach, which also works:

```java
    Metrics metrics = new Metrics();
    metrics.addMetric("test.name", new Total());
    metrics.metrics().values(); // and here we finally got it :)
```

I've documented the above mentioned approach in javadocs for `KafkaMetric` in scope of this PR.

stepio replied:
If you think that such an implementation may be useful for others, I'll add it to `org.apache.kafka.test`. I'm not sure that method `of` (or public constructor) is suitable in `KafkaMetric` itself - I'm afraid they could be misused in main code.

guozhangwang replied:
I see, maybe we can then remove this class and the corresponding test case, and just mention that you can also extend this class as a mock yourself for testing purposes in the same java doc paragraph?

guozhangwang commented on clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:
```diff
@@ -238,6 +278,8 @@ public AttributeList setAttributes(AttributeList list) {
             throw new UnsupportedOperationException("Set not allowed.");
         }
 
+        public int size() {
```
nit: could we rename it to sth. like "numAttributes"?

guozhangwang commented on clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:
```diff
@@ -63,69 +64,74 @@ public void configure(Map<String, ?> configs) {}
-            for (KafkaMbean mbean : mbeans.values())
+            for (KafkaMetric metric : metrics) {
+                final KafkaMbean mbean = addAttribute(metric);
+                if (!this.mbeans.containsKey(mbean.metricName)) {
```
Since `addAttribute` always calls `getMBean`, `this.mbeans` should always contain this metric already after `addAttribute`, right? Ditto below at line 83.

stepio replied:
I've moved all updates for a non-thread-safe `mbeans` map to `synchronized` blocks. So now neither `addAttribute`, nor `getMBean` updates it. In my opinion this improves readability.

guozhangwang replied:
I see. That makes sense.

guozhangwang commented on clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:
```diff
@@ -63,69 +64,74 @@ public void configure(Map<String, ?> configs) {}
-        KafkaMbean mbean = this.mbeans.get(mBeanName);
-        if (mbean != null)
+        final MetricName metricName = metric.metricName();
+        final KafkaMbean mbean = getMBean(metricName);
```
We are slightly changing the behavior here: if the metric does not exist, previously it is just an no-op, while now we will create the metric first then deletes it.

stepio replied:
I know... that's a questionable improvement, I agree. The idea was to use a common accessor in all the places. If you think that overhead is too much, please inform me - I'll revert the change.

P.S.: Also this update allows removing extra `null` check in `metricRemoval` method.

guozhangwang replied:
On a second thought, the overhead should be minimal: a few young gen objects only. So mvn.

ijuma replied:
We've had a report of large amounts of memory being used by `KafkaMbean` with empty maps. This makes it worse, so I don't think we should do it.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:
```diff
@@ -313,16 +318,16 @@ public boolean conditionMet() {
                         return false;
                     } catch (InvalidStateStoreException e) {
-                        // there must have been at least one rebalance state
+                        log.debug("There must have been at least one rebalance state", e);
```
This message is OK as a comment, but may be misleading as an debug log entry. If you do want to log it, maybe `The state store may has been closed already due to consecutive rebalances` is better. Ditto below.

guozhangwang reviewed (commented):
Left a few comments.

guozhangwang commented:
About `assertj`, seems it does not add more asserting functionalities that junit cannot provide? Is there any specific reasons you want to add a new dependency jar into Kafka?

ijuma reviewed (commented):
Thanks for the PR @stepio. To make it easier to review (and eventually merge), can you edit the PR title and description with a summary of what the PR is actually doing now (since we decided not to change the `MetricsReporter` interface for now).

stepio commented:
> Is there any specific reasons you want to add a new dependency jar into Kafka?

It's a test-dependency, which incorporates functionality of `junit` and `hamcrest-matchers`. As for me, I love the "builder-like" syntax for writing tests - allows applying multiple checks for a single value. Also no mess with "actual" and "expected" values (which one is the first, and which one is a second). As a bonus, it supports Java 8 constructions starting from 3.x.x version.

But it's still just a suggestion.

guozhangwang commented:
> But it's still just a suggestion.

Personally I'd try to avoid growing dependencies, both runtime and test unless it is necessary. For `assertj` I found it provides some syntax sugar and more clear argument semantics but not providing any asserting functionalities that `junit` and `harmcrest` already provides.

stepio commented:
Hi @guozhangwang  and @ijuma 

Did more updates, so it's your turn again :)

guozhangwang commented:
LGTM overall. Could you respond to @ijuma 's comment?

stepio commented:
> LGTM overall. Could you respond to @ijuma 's comment?

Thanks, @guozhangwang ! I was unable to respond here, so I went to "FIles changed" tab of the pull request and replied there:
<img width="780" alt="screenshot 2017-01-06 08 19 11" src="https://cloud.githubusercontent.com/assets/5202414/21709357/14aa1e0a-d3e9-11e6-8976-7997e34effc7.png">

Unfortunately that answer is not displayed here. @ijuma, is it visible to you in the diff?

stepio commented:
Hi @ijuma,

Any update? I've replied to your comment, but have not changed anything, as I haven't seen any clear suggestion... So what do you think?

stepio committed:
KAFKA-3923: Update unit tests, re-factor JmxReporter

ijuma commented on clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:
```diff
@@ -63,91 +64,90 @@ public void configure(Map<String, ?> configs) {}
-            MetricName metricName = metric.metricName();
-            String mBeanName = getMBeanName(metricName);
-            if (!this.mbeans.containsKey(mBeanName))
-                mbeans.put(mBeanName, new KafkaMbean(mBeanName));
```
What is the benefit of moving this out of this method? It seems like we end up having to do it in the 2 callers?

stepio replied:
While updating unit tests (as requested by @guozhangwang) and reading this code I was puzzled much with the fact that non-thread-safe `HashMap` is modified in this method. Then I've found that the method is called only from a synchronized block, so it's safe. But still to improve readability and help future maintainers I've decided to move all such modifications to the appropriate synchronized blocks and keep helper methods just for "routine".

ijuma replied:
I think this change is debatable. You make it a bit more visible that the lock is held when mutating the structure, but some of the logic is duplicated. I'd leave it as it was, personally.

ijuma commented on clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:
```diff
@@ -160,17 +160,30 @@ private void reregister(KafkaMbean mbean) {
     }
 
-    private static class KafkaMbean implements DynamicMBean {
+    public int mbeansQty() {
```
This method name doesn't follow Kafka code conventions.

stepio replied:
Maybe. Where can I see the above mentioned code conventions? Where is the violation?
Tried googling and found next guideline:
https://kafka.apache.org/coding-guide.html
Don't see the exact point which I've violated.

Tried searching the code  - don't see much `Qty`, `Num`, or `Size` appendixes, but found some `Count`... Is this what you've meant?

ijuma replied:
Yes, we don't use `Qty` as a suffix. We sometimes use `num` prefix and other times `Count` suffix (it would be nice to be consistent, but we're not there yet).

ijuma commented on clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:
```diff
@@ -204,18 +219,10 @@ public KafkaMetric removeAttribute(String name) {
-                                                  false,
-                                                  false);
-                i += 1;
+                attrs[i++] = new MBeanAttributeInfo(entry.getKey(), double.class.getName(), entry.getValue().metricName().description(), true, false, false);
```
Nit: this line is too long.

ijuma commented on clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:
```diff
@@ -160,17 +160,30 @@ private void reregister(KafkaMbean mbean) {
+            }
+        }
+
+        private KafkaMbean(ObjectName objectName) throws MalformedObjectNameException {
```
This exception is no longer possible since the constructor is taking `ObjectName`.

ijuma commented on clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:
```diff
@@ -238,6 +245,40 @@ public AttributeList setAttributes(AttributeList list) {
+         * @param metricName identifier for {@link KafkaMetric}
+         * @return standard JMX MBean name in the following format domainName:type=metricType,key1=val1,key2=val2
+         */
+        protected static ObjectName getObjectName(MetricName metricName, String prefix) {
```
We typically avoid the `get` prefix in Kafka, so `objectName` seems fine.

ijuma commented on clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java:
```diff
@@ -16,25 +16,59 @@
+            reporter.getMBean(null);
+        } catch (RuntimeException e) {
+            assertThat(e, CoreMatchers.instanceOf(NullPointerException.class));
+            assertThat(e.getMessage(), CoreMatchers.is("Parameter 'metricName' is mandatory"));
```
If using hamcrest matchers, I think it's better to use static imports to make them more concise.

ijuma reviewed (commented):
Sorry for the delay, left a few comments.

stepio commented:
@ijuma 
Thanks for the detailed review. Seems that it might be reasonable to reconsider not only the exact changes, but the pull request in general.

guozhangwang commented:
About the duplicated code: I know it is subjective, but to me I feel this is worthwhile to make the locking scope clearer. And it adds about 6 lines only.

guozhangwang commented:
Closing for cleanup.
